test/e2e/cluster_samples_operator_test.go

```diff
@@ -938,3 +941,89 @@ func TestRecreateDeletedManagedSample(t *testing.T) {
 	verifyDeletedTemplatesRecreated(t)
 	t.Logf("Config after TestRecreateDeletedManagedSample: %#v", verifyOperatorUp(t))
 }
+
+func TestUpgrade(t *testing.T) {
+	verifyOperatorUp(t)
+	err := verifyConditionsCompleteSamplesAdded(t)
+	if err != nil {
+		dumpPod(t)
+		t.Fatalf("samples not stable at start of upgrade test %#v", verifyOperatorUp(t))
+	}
+
+	newVersion := fmt.Sprintf("%v", rand.Intn(10000000))
+
+	// update env to trigger upgrade
+	depClient := kubeClient.AppsV1().Deployments("openshift-cluster-samples-operator")
+	err = wait.PollImmediate(1*time.Second, 1*time.Minute, func() (bool, error) {
+		dep, err := depClient.Get("cluster-samples-operator", metav1.GetOptions{})
+		if err != nil {
+			t.Logf("error waiting for operator deployment to exist: %v\n", err)
+			return false, nil
+		}
+		t.Logf("found operator deployment")
+		for i, env := range dep.Spec.Template.Spec.Containers[0].Env {
+			t.Logf("looking at env %s", env.Name)
+			if strings.TrimSpace(env.Name) == "RELEASE_VERSION" {
+				t.Log("updating RELEASE_VERSION env")
+				dep.Spec.Template.Spec.Containers[0].Env[i].Value = newVersion
+				_, err := depClient.Update(dep)
+				if err == nil {
+					return true, nil
+				}
+				t.Logf("%v", err)
+				if kerrors.IsConflict(err) {
+					return false, nil
+				}
+				if stub.IsRetryableAPIError(err) {
+					return false, nil
+				}
+				return false, err
+			}
+
+		}
+		return false, nil
+	})
+	if err != nil {
+		dumpPod(t)
+		t.Fatalf("problem updating deployment env")
+	}
+
+	err = wait.PollImmediate(1*time.Second, 1*time.Minute, func() (bool, error) {
+		cfg := verifyOperatorUp(t)
+		if cfg.ConditionTrue(samplesapi.MigrationInProgress) {
```

Comment:
this seems like a timing window based test (there's no guarantee you manage to observe the operator while it's migrating things)

Reply:
disagree on the "no guarantee you can manage to observe" part 

the samples operator's config object is not going away

whenever the pod does recycle, and enters the migration path, it is suppose to set this condition on the config object

can you elaborate on this concern, or maybe describe it with a specific set of events from the scenario?

Reply:
What if migration completes (thus goes back to false) before you enter this polling loop?  Or it goes false/true/false in less than your polling interval so you never see it?

Reply:
that is soooo unlikely given the particulars of the system and this test 

we start checking as soon as we update the env, and I've never seen an env induced redeployment happen in milliseconds

 it does not seem worth adding additional complexity to account for that

and if we never see it going to true, it more than likely points to something wrong underneath us that we want to investigate

Reply:
i wasn't thinking of adding complexity, I was thinking of just removing this check since you're ensuring migration happened in other ways anyway.

but if you're confident it won't flake, leave it I guess...  i just get nervous any time i see timing-dependent assumptions in tests.

Reply:
I see 

understood, but yeah, seeing that config obj condition change is important in my mind vs. just checking the version, as it narrows the focus of the diagnosis if a failure did occur

so gonna leave it ... if it bites me later on, you know what you can say :-)

thanks

Reply:
I hope to be blissfully uninvolved if/when that happens :)